lib/widgets/emoji_reaction.dart

```diff
@@ -360,3 +366,261 @@ class _TextEmoji extends StatelessWidget {
+    final designVariables = DesignVariables.of(context);
+
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      children: [
+        Padding(padding: const EdgeInsetsDirectional.only(start: 8, top: 4),
+          child: Row(children: [
+            // TODO(design): Make sure if we need a button to clear the textfield.
+            Flexible(child: Padding(
+              padding: const EdgeInsets.symmetric(vertical: 2),
+              child: TextField(
+                controller: _controller,
+                autofocus: true,
+                decoration: InputDecoration(
+                  hintText: zulipLocalizations.emojiPickerSearchEmoji,
+                  contentPadding: const EdgeInsetsDirectional.only(start: 10, top: 6),
+                  filled: true,
+                  fillColor: designVariables.bgSearchInput,
+                  border: OutlineInputBorder(
+                    borderRadius: BorderRadius.circular(10),
+                    borderSide: BorderSide.none),
+                  hintStyle: TextStyle(color: designVariables.textMessage)),
+                style: const TextStyle(fontSize: 19, height: 26 / 19)))),
+            TextButton(
+              onPressed: () => Navigator.pop(context),
+              style: TextButton.styleFrom(
+                padding: EdgeInsets.zero,
+                splashFactory: NoSplash.splashFactory,
+                foregroundColor: designVariables.contextMenuItemText,
+              ).copyWith(backgroundColor: WidgetStateColor.resolveWith((states) =>
+                states.contains(WidgetState.pressed)
+                  ? designVariables.contextMenuItemBg.withFadedAlpha(0.20)
+                  : Colors.transparent)),
+              child: Text(zulipLocalizations.dialogClose,
+                style: const TextStyle(fontSize: 20, height: 30 / 20))),
+          ])),
+        Expanded(child: InsetShadowBox(
+          top: 8, bottom: 8,
+          color: designVariables.bgContextMenu,
+          child: ListView.builder(
+            padding: const EdgeInsets.symmetric(vertical: 8),
+            itemCount: _resultsToDisplay.length,
+            itemBuilder: (context, i) => EmojiPickerListEntry(
+              pageContext: widget.pageContext,
+              emoji: _resultsToDisplay[i].candidate,
+              message: widget.message)))),
+    ]);
+  }
+}
+
+@visibleForTesting
+class EmojiPickerListEntry extends StatelessWidget {
+  const EmojiPickerListEntry({
+    super.key,
+    required this.pageContext,
+    required this.emoji,
+    required this.message,
+  });
+
+  final BuildContext pageContext;
+  final EmojiCandidate emoji;
+  final Message message;
+
+  static const _emojiSize = 24.0;
+  static const _notoColorEmojiTextSize = 20.1;
+
+  void _onPressed() {
+    // Dismiss the enclosing action sheet immediately,
+    // for swift UI feedback that the user's selection was received.
+    Navigator.pop(pageContext);
+
+    doAddOrRemoveReaction(
+      context: pageContext,
```

Comment:
I see, this is why we end up needing to pass `pageContext` down through all these layers (rather than each widget using its own BuildContext like usual). It's so that if the add/remove request fails, `doAddOrRemoveReaction` has a (probably-)still-mounted context it can use for showing the error.

I feel like there's probably some better pattern we could use here. But I'm not sure offhand what it'd look like — will probably take a bit of experimenting — and this is the natural extension of the pattern we already use throughout the action sheets. So this pattern is good to stick with for this PR.